### DISCUSSION TITLE
Network communication using OSC

### DISCUSSION THREAD

**Discussion by glopesdev** [bonsai-rx/bonsai#872] [Articles]
Proposed outline:
- Show how to configure connections
- Show how to parse and format messages
- Show examples of client server pattern
- Show how you can use that to talk to Python / MATLAB

**Comment by horsto** [comment]
This is great - I would highly appreciate some help with this. I stumbled over https://groups.google.com/g/bonsai-users/c/zXS_1MF8nxM/m/DnkJ5UxkAwAJ and similar discussions about OSC (specifically in combination with Matlab) in the google group. 

The trial structure of my experiment is determined in Matlab, and I want functions to trigger Bonsai / enable real time communication through OSC. 

Here is what I want to achieve: 
- I want to use the Bonsai CLI (https://github.com/bruno-f-cruz/Bonsai-CLI-Guide) to start a Bonsai workflow from a Matlab function 
- Once started I want this workflow to hold execution until a file name (external property) has been communicated via OSC and a "start", or similar, has been sent (all through Matlab). This will start the acquisition of camera frames / display online tracking results.
- I want to stop the execution of the workflow upon sending "stop" via OSC, or, alternatively, somehow stop Bonsai altogether through CLI (I did not see that this is possible).

I managed to send commands via OSC from Matlab to Bonsai (that is nice!), however, many elements here are missing, e.g. how to control the execution of the main workflow through OSC and, crucially, how to stop it. 

It would be great to have a template here for how to do this in Bonsai (I just started using Bonsai). This design pattern might be useful for others too.

**Comment by spencerblackwood** [comment]
I use OSC to communicate between Bonsai and python using the python-osc module. I just uploaded a simple example showing how I read EEG data coming from Bonsai into a Jupyter notebook then send integer messages back to Bonsai. Hopefully that will help someone.

Code is here:
https://github.com/spencerblackwood/bonsai-osc-jupyter-example

**Reply by jvoigts** [reply]
In case it's useful for others in my situation, I just uploaded a similar example here: https://github.com/jvoigts/bonsai-osc-bidirectional-example the main difference is probably that my implementation is using a blocking Osc server on the python side so this can work inside an interactive user interface.
